lib/ansible/modules/windows/win_psmodule.ps1

```diff
@@ -1,59 +1,317 @@
+        }
+        elseif ( $RequiredVersion ) {
+            $ReturnedModule = $ExistingModules | Where-Object -FilterScript { $_.FullVersion -eq $RequiredVersion }
+        }
+        elseif ( $MinimumVersion -and $MaximumVersion ) {
+            $ReturnedModule = $ExistingModules | Where-Object -FilterScript { $MinimumVersion -le $_.Version -and $MaximumVersion -ge $_.Version } | Select-Object -First 1
+        }
+        elseif ( $MinimumVersion ) {
+            $ReturnedModule = $ExistingModules | Where-Object -FilterScript { $MinimumVersion -le $_.Version } | Select-Object -First 1
+        }
+        elseif ( $MaximumVersion ) {
+            $ReturnedModule = $ExistingModules | Where-Object -FilterScript { $MaximumVersion -ge $_.Version } | Select-Object -First 1
+        }
+    }
+
+    $ReturnedModuleCount = ($ReturnedModule | Measure-Object).Count
+
+    if ( $ReturnedModuleCount -eq 1 ) {
+        $ExistingModule.Exists = $true
+        $ExistingModule.Version = $ReturnedModule.FullVersion
+    }
+
+    $ExistingModule
+}
+
+Function Add-DefinedParameter {
+    Param (
+        [Parameter(Mandatory=$true)]
+        [Hashtable]$Hashtable,
+        [Parameter(Mandatory=$true)]
+        [String[]]$ParametersNames
+    )
+
+    ForEach ($ParameterName in $ParametersNames) {
+        $ParameterVariable = Get-Variable -Name $ParameterName -ErrorAction Ignore
+        if ( $ParameterVariable.Value -and $Hashtable.Keys -notcontains $ParameterName ){
+                $Hashtable.Add($ParameterName,$ParameterVariable.Value)
+        }
+    }
+
+    $Hashtable
+}
+
+Function Install-PsModule {
+    Param(
+        [Parameter(Mandatory=$true)]
+        [String]$Name,
+        [String]$RequiredVersion,
+        [String]$MinimumVersion,
+        [String]$MaximumVersion,
+        [String]$Repository,
+        [Bool]$AllowClobber,
+        [Bool]$SkipPublisherCheck,
+        [Bool]$AllowPrerelease,
+        [Bool]$CheckMode
+    )
+
+    $ExistingModuleBefore = Get-PsModule -Name $Name -RequiredVersion $RequiredVersion -MinimumVersion $MinimumVersion -MaximumVersion $MaximumVersion
+
+    if ( -not $ExistingModuleBefore.Exists ) {
+        try {
+            # Install NuGet provider if needed.
+            Install-NugetProvider -CheckMode $CheckMode
+
+            $ht = @{
+                Name = $Name
+                WhatIf = $CheckMode
+                Force = $true
+            }
+
+            [String[]]$ParametersNames = @("RequiredVersion","MinimumVersion","MaximumVersion","AllowPrerelease","AllowClobber","SkipPublisherCheck","Repository")
+
+            $ht = Add-DefinedParameter -Hashtable $ht -ParametersNames $ParametersNames
+
+            Install-Module @ht -ErrorVariable ErrorDetails | out-null
+
+            $result.changed = $true
+            $result.output = "Module $($Name) installed"
+        }
+        catch {
+
+            if ( $ErrorDetails.Exception.Message ) {
```

Comment:
This whole section doesn't seem like it should be needed. We should also stop having generic try/catch blocks in our code. In the past Ansible wasn't the best at handling errors but in devel/2.8 this isn't required. All we are doing is just obfuscating the error/exception by doing this.

Reply:
I removed that section.

I agree that error handling can be made better by using new capabilities of Ansible 2.8. I'll try to do that later. For now, I don't have spare time. Sorry.

Reply:
I don't understand why you can't do that in this PR. It's as simple as removing the try/catch blocks that do nothing but say `Failed to do x: $($_.Exception.Message)`. You don't have to add code, just remove it and make sure the indentation is correct.

Reply:
Done.